### PR TITLE
Extrapolating the bcc address from the envelope

### PR DESCRIPTION
**Pull request by hgcode**
Will create a pull request on Griddler now that has minimal changes and depends on this parser to actually get the bcc. Hopefully this will help.

In response to this: 
https://github.com/thoughtbot/griddler/issues/168

### CODE REVIEW

**Comment by bbeaird** [comment]
From reading issues and pull requests about the griddler and griddler-sendgrid gems, it seems like this the limiting factor on getting access to the bcc attribute, yea? Or am I missing something? Thx.

**Comment by gabebw** [comment]
Yes, after this is merged, BCC support should be available.

@hgcode Can you take a look at our feedback?

**Comment by bbeaird** [comment]
@gabebw - FYI - I ended up using hgcode's version of the gem. It works. While it might not be up to Thoughtbot's code standards ;0), it might be worth merging in anyways given, IMHO, the usefulness of being able to access the bcc attribute.

PS - Parsing to, cc, and bcc is a bitch, and it's easy to get duplicates. In my limited experience, the best way is to use only the bcc field when parsing incoming emails. For example, the following email...

to: a@123.com, b@123.com
cc: c@123.com, d@123.com
bcc: e@123.com, f@123.com

actually sends 3 emails with the following fields.

Email #1 
to: a@123.com, b@123.com
cc: c@123.com, d@123.com
bcc: e@123.com

Email #2 
to: a@123.com, b@123.com
cc: c@123.com, d@123.com
bcc: f@123.com

Email #3 
to: a@123.com, b@123.com
cc: c@123.com, d@123.com
bcc: a@123.com, b@123.com, c@123.com, d@123.com

That last one is the one that threw me for a loop. If you only look at the bcc field, you'll end up with the right emails and no duplicates.

**Comment by hgcode** [comment]
hey guys sorry I've been swamped with RL.. will refactor this eventually and do a pull request

**Comment by bbeaird** [comment]
Hi @hgcode - first off, thanks for your code. It seems to work great in 99% of cases; however, I may have run across a bug. My app is an email reminder service. You can send emails to things like 2weeks@ndgml.com, 1minute@ndgml.com, etc. People will use the to, cc, and bcc fields in all sorts of ways. 

Currently, I'm using your created bcc attribute in the following way to create reminders.

    @bcc = email.bcc
    def set_to_cc_and_bcc_fields
      if @bcc.present?
        @to_cc_and_bcc = @bcc
      else
        @to_cc_and_bcc = (@to + @cc).uniq
      end
    end

Since bcc'd emails get delivered as separate emails, doing the above avoids the creation of duplicate nudges (reminders). The way bcc'd emails are delivered threw me for a loop. 

Anyways, the inconsistency that I'm observing is for emails is when using "full email addresses" like ```friday <friday@ndgml.com>``` compared to "non-full email addresses" like ```1m@ndgml.com```. The two example below illustrate this. In the first example, both emails are "non-full" email addresses. Nothing shows up in the bcc field, this makes sense. In example 2, one email address is full and the other is non-full. The full email address shows up in the bcc attribute. This is throwing off my set_to_cc_and_bcc_fields method.  

Example 1
```
App 15107 stdout: "email.to!!!!!!!!!!!"
App 15107 stdout: [{:token=>"1m", :host=>"ndgml.com", :email=>"1m@ndgml.com", :full=>"1m@ndgml.com", :name=>nil}]
App 15107 stdout: "email.cc!!!!!!!!!!!!"
App 15107 stdout: [{:token=>"thursday", :host=>"ndgml.com", :email=>"thursday@ndgml.com", :full=>"thursday@ndgml.com", :name=>nil}]
App 15107 stdout: "email.bcc!!!!!!!!!!!!"
App 15107 stdout: []
```

Example 2 
```
App 15107 stdout: "email.to!!!!!!!!!!!"
App 15107 stdout: [{:token=>"1m", :host=>"ndgml.com", :email=>"1m@ndgml.com", :full=>"1m@ndgml.com", :name=>nil}]
App 15107 stdout: "email.cc!!!!!!!!!!!!"
App 15107 stdout: [{:token=>"friday", :host=>"ndgml.com", :email=>"friday@ndgml.com", :full=>"friday <friday@ndgml.com>", :name=>"friday"}]
App 15107 stdout: "email.bcc!!!!!!!!!!!!"
App 15107 stdout: [{:token=>"friday", :host=>"ndgml.com", :email=>"friday@ndgml.com", :full=>"friday@ndgml.com", :name=>nil}]
```

I've gone through your code, and can't figure out why this would be happening. I know your busy, but any insights from a pro like your would be greatly appreciated. Thanks!

PS - What is RL?

**Comment by kevindewalt** [comment]
Agree with @bbeaird. Using @hgcode fork, bcc working fine.

Lots of use cases for bcc, suggest merging.

File: lib/griddler/sendgrid/adapter.rb [hunk 1]
```diff
@@ -26,6 +27,33 @@ def recipients(key)
         ( params[key] || '' ).split(',')
       end
 
+      def get_bcc
+        return [] unless params[:envelope].present?
+        envelope = json_to_hash(params[:envelope])
+        bcc = if envelope.present?
```
Comment (kelvinst) [inline lib/griddler/sendgrid/adapter.rb]: @gabebw I prefer this way:

```ruby
bcc = envelope["to"] || []
```

IMHO, looks more readable than the `.fetch` method.

**Comment by plicjo** [comment]
Will you guys take a look at the code now? I implemented your suggestions where I could.

Unfortunately, quite a few of the suggestions broke the test suite.

**Comment by hgcode** [comment]
I think that should be fixed though, they are not going to merge if there are any failing tests...

**Comment by calebthompson** [comment]
@wingrunr21

**Comment by plicjo** [comment]
All the tests pass currently, @hgcode!

**Comment by hgcode** [comment]
Cool! thanks

**Comment by wingrunr21** [comment]
@calebthompson What am I looking at here?

**Comment by calebthompson** [comment]
I was under the impression that you'd taken over this as well, but hadn't gotten access for some reason. If that's not right no biggie just let me know.

**Comment by wingrunr21** [comment]
No, I thought it was just griddler. I don't mind maintaining this but as I don't actually use SendGrid I'm not sure I'm best for this gem.

**Comment by calebthompson** [comment]
Okay cool. I'll merge this then.

**Comment by plicjo** [comment]
@bbeaird had some legitimate concerns, guys!

**Comment by calebthompson** [comment]
Ah, it seemed like the conversation had settled that those had been addressed, sorry. Yeah a separate issue would be good. My guess is still that those are BCC not being sent since it's not required to be sent (and generally isn't in my experience). Any mail server in the line could potentially remove it I think.

```
The "Bcc:" field (where the "Bcc" means "Blind Carbon Copy") contains
   addresses of recipients of the message whose addresses are not to be
   revealed to other recipients of the message.  There are three ways in
   which the "Bcc:" field is used.  In the first case, when a message
   containing a "Bcc:" field is prepared to be sent, the "Bcc:" line is
   removed even though all of the recipients (including those specified
   in the "Bcc:" field) are sent a copy of the message.  In the second
   case, recipients specified in the "To:" and "Cc:" lines each are sent
   a copy of the message with the "Bcc:" line removed as above, but the
   recipients on the "Bcc:" line get a separate copy of the message
   containing a "Bcc:" line.  (When there are multiple recipient
   addresses in the "Bcc:" field, some implementations actually send a
   separate copy of the message to each recipient with a "Bcc:"
   containing only the address of that particular recipient.) Finally,
   since a "Bcc:" field may contain no addresses, a "Bcc:" field can be
   sent without any addresses indicating to the recipients that blind
   copies were sent to someone.  Which method to use with "Bcc:" fields
   is implementation dependent, but refer to the "Security
   Considerations" section of this document for a discussion of each.
```

**Comment by calebthompson** [comment]
https://tools.ietf.org/html/rfc2822

**Pull request merged**